### PR TITLE
Unit 2 Sprint 6

### PR DESCRIPTION
**Pull request by ndoshi83**

### CODE REVIEW

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# making_forecasts - "Nayan Doshi"

# Part 1 - API Load
***
**What went well:**

- Great job using datetime! 

```python

df['Date'] = pd.to_datetime(df['Date'])
df = df.set_index('Date')


```

- Clean and simplisitic code with a nice graph:
```python

df.plot(figsize=(30,10))

```

-Nice use of Regex and datetime:
```python
df['timestamp'] = pd.to_datetime(df['timestamp'],format="%Y%m%d%H")
df.head()
```
**Where you could of improved on:**

-I'd enlarge the X asis and Y axis font and set a title :

```python
https://plot.ly/matplotlib/figure-labels/
```


_____________________
# Quick View

***

**What went great:**

-Good job setting figure size:

```python

df.plot(figsize=(16,9))```



**-Great use of reindex :**
```python

df = df.reindex(pd.date_range(start='2015-07-01',end='2020-05-24'))

```


**Where you could of improved on:**

- Using crosstab you can achieve powerful things, combined with loops and chi-squared tests:

```python

```

# Part 3 -- Average method:

- Great job with this first assignment, and keep up the exemplary work:
```python
df['mean'] = df['views'].mean()
```

# What to work on
- [x] Practice formatting and regex
- [x] Comment more

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
#Nayan Doshi- "Regression 1 Sprint Challenge"

# Part 1. Predict avocado sales
***
**What went well:**

-Good job imporing data and dataset and checking out shape:

```python
:
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)
print(avocado.shape)


```

## 1.2. Plot the forecast

- Impressive use of matplotlib ( showed example of using Prohet below):
```python



In [12]:

# Plot forecast
fig1 = m.plot(forecast)




```

- Using Prophet: 

```python


from fbprophet import Prophet
avocado.columns = ['ds', 'y']
m = Prophet()
m.fit(avocado)
future = m.make_future_dataframe(periods=365*2)
forecast = m.predict(future)


fig1 = m.plot(forecast)


```




```

##1.5. Plot the cross-validation Mean Absolute Percentage Error: metric='mape' 
***
**What went well:**

- Clean and elegant :

```python


fig1 = m.plot(forecast)

```


** What could be improved: 

```python 
from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(df_cv, metric='mape')

#This is what we use 
```

#Stretch Goal (PARTIAL)

# Part 2 

##2.1. Begin with baselines for regression. Calculate the mean absolute error and 𝑅2 score for a mean baseline.

- Went went the base .head: 


``` python


rent.head()





```

**What could be better

-use this syntax for better insight:
```python

rent.describe()

```


## 2.2. Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
-Good job choosing features and getting to it (I added another way as well)):
```python

#Your Code
X = rent[['bedrooms','bathrooms']]
y = rent['price']


features = ['bedrooms','bathrooms']

target = 'price'

X = rent[features]
y = rent[target]

model.fit(X, y)

y_pred = model.predict(X)



train = rent.query('month <= 5')
test  = rent.query('month == 6')
X_train = train[features] 
y_train = train[target]
X_test  = test[features]
y_test  = test[target]


model.fit(X_train, y_train)


y_pred = model.predict(X_test)



#My Syntax

features = ['bedrooms']
target = 'price'
x = rent[features]
y = rent[target]


## Other option

train = rent.query('month < 6')
test = rent.query('month == 6')
train.shape, test.shape

features = ['bedrooms', 'bathrooms']
target = 'price'

```

##2.3. Choose any two features.

-use this syntax for better insight:
```python

```

2.4. Use scikit-learn to fit a Linear Regression model on the train data.

-Not bad for just starting out to use to scikit. Syntax is not bad, but features could be better ( see below ):
```python

from sklearn.metrics import mean_absolute_error, r2_score
import matplotlib.pyplot as plt

def plot_preds(x, y, y_pred):
    plt.scatter(x, y, label='y_true')
    plt.plot(x, y_pred, label='y_pred')
    plt.legend()
    
    
    mae = mean_absolute_error(y, y_pred)
    r2 = r2_score(y, y_pred)
    print('Mean Absolute Error:', mae)
    print('R^2:', r2)
    
    
    
plot_preds(x, y, y_pred)
plt.title('Mean Baseline');

```

**What could be better

-better syntax:


```python

# Choose a class of model by importing the appropriate estimator class from Scikit-Learn.
from sklearn.linear_model import LinearRegression

# Choose model hyperparameters by instantiating this class with desired values.
model = LinearRegression()

# Arrange data into a features matrix and target vector.
X_train = train[features]
y_train = train[target]
X_test = test[features]
y_test = test[target]

X_train.shape, y_train.shape, X_test.shape, y_test.shape

# Fit the model to your data by calling the fit() method of the model instance.
model.fit(X_train, y_train)

# Apply the Model to new data: For supervised learning, often we predict labels for unknown data using the predict() method.
y_pred = model.predict(X_test)

```

2.6. Get regression metrics MAE and 𝑅2 for the test data.

-Great job:
```python


mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print('Test Mean Absolute Error:', mae)
print('Test R^2 Score:', r2)

```

2.7. Get the model's coefficients and intercept.

https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html

-use this syntax for better insight:
```python
model.intercept_, model.coef_
```

#Stretch Goals_2

```python

import seaborn as sns

target = 'price'
features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
for feature in features:
    sns.lmplot(x=feature, y=target, data=train, scatter_kws={'alpha':0.01})
    
    ```
    
    
    
    ```python
    
    
    def fit(train, test, features):
    X_train = train[features]
    y_train = train[target]
    X_test = test[features]
    y_test = test[target]

    model = LinearRegression()
    model.fit(X_train, y_train)

    print(f'Linear Regression with {len(features)} features: {features}')
    print('Train Mean Absolute Error:', 
          mean_absolute_error(y_train, model.predict(X_train)))
    print('Train R^2 Score:', 
          r2_score(y_train, model.predict(X_train)))
    print('Test Mean Absolute Error:', 
          mean_absolute_error(y_test, model.predict(X_test)))
    print('Test R^2 Score:', 
          r2_score(y_test, model.predict(X_test)))

features = ['bedrooms', 'bathrooms', 'latitude', 'longitude']
fit(train, test, features)
```

-with 3 features:

```python
features = ['bedrooms', 'bathrooms', 'longitude']
fit(train, test, features)
```

# What to work on
- [x] Comment more in code 
- [x] Less lines and more organized notebooks 
- [x] Come to PM for questions or help

**Review by brennerhaverlock — CHANGES_REQUESTED** [review]
# Nayan Nicolae- Regression 1 Sprint Challenge

# Part 1. Predict avocado sales 

***
**What went well:**

-Good collective use of imports and reading in data of the :

```python
import pandas as pd
avocado_url = 'https://drive.google.com/uc?export=download&id=1ljE-LhCA9CCFvSIJYJ1ewX8JmdDpfuhi'
avocado = pd.read_csv(avocado_url)
assert avocado.shape == (169, 2)
print(avocado.shape)



```
#Part 2. Predict NYC apartment rent 


## Imports

- Nice job reading from Google Drive and finding shape data:

```python

import pandas as pd
rent_url = 'https://drive.google.com/uc?export=download&id=1lVVPmh-WYGb_0Gs_lOv22i02-150qDdZ'
rent = pd.read_csv(rent_url)
assert rent.shape == (48295, 7)
print(rent.shape)


```


## Prophet Plotting 

- Very simple but powerful code! Nice use of comments too :

```python


# Plot forecast
fig1 = m.plot(forecast)




# Plot components
fig2 = m.plot_components(forecast)




```

- Good use of [cross_validation](https://www.google.com) with Facebook's diagnostics API: 

```python

# Create time series cross-validation using the Prophet functionality
# Import needed functions
from fbprophet.diagnostics import cross_validation
avocado_cv = cross_validation(avocado_model, initial='730 days', 
                         horizon='90 days', period='45 days')
```

## Calculating MAPE(Mean Absolute Percentage Error)

- Impressive use of finding the MAPE!: 

```python
# Use Prophet to plot MAPE
from fbprophet.plot import plot_cross_validation_metric
fig = plot_cross_validation_metric(avocado_cv, metric='mape')

```
# Part 2. Predict NYC apartment rent

## Setting Target

- Nice job setting feature and target of rent:  

```python
# Selecting 2 features to use for linear regression testing
features = ['bedrooms', 'bathrooms']

target = 'price'
```

## MAE

- Excellent work. I mean this because this is important when looking at features:

```python
#Calculate regression metrics for predicted targets on test data
rent_test_mae = mean_absolute_error(rent_test[target],y_pred)
rent_test_r2 = r2_score(rent_test[target],y_pred)

print('The MAE for test is:', rent_test_mae, '\n')
print('The R^2 for test is:', rent_test_r2)

```
#What to work on
- [x] Comment more in code
- [x] Come to PM for questions or help
